Add Elasticsearch-hadoop pushdown optimizer support in hive module

AbelChan1989 opened the pull request:
HI,ALL

  the latest version of Elasticsearch-hadoop does not support the pushdown optimization in hive module, the PR is mainly to add this support.

  Implementation plan:
  	1.Implements the HiveStoragePredicateHandler interface on EsStorageHandler and uses it as a pushdown optimizer.
  	2.The HiveStoragePredicateHandler is as a pushdown optimization implementation class, which is convert the ExprNodeDesc from hive to the corresponding Elasticsearch query.
  	3.The SargableParser is as a pushdown optimization conversion class, which is to convert hive operator grammar to the corresponding Elasticsearch syntax.
  	4.the result of pushdown optimization merges to the es.query property, and starts the next step: Elasticsearch Scroll Operation.

AbelChan1989 commented:
In addition, the parameters which can support user-defined as follows:
	1.Is pushdown optimization function that the name is 'es.internal.hive.pushdown' and the default is true.
	2.Pushdown optimization implementation class which is need to inherit org.elasticsearch.hadoop.hive.pushdown.PredicateHandler class, the default implementation class: org.elasticsearch.hadoop.hive.EsStoragePredicateHandler
	3.Pushdown optimization conversion class, the need to implement org.elasticsearch.hadoop.hive.pushdown.SargableParser interface, the default implementation class: Org.elasticsearch.hadoop.hive.pushdown.EsSargableParser

jbaiera commented:
@AbelChan1989 This is a really impressive (and large, wow) PR. I will start taking a look at it review wise tomorrow morning, but until then, really awesome stuff so far.

jbaiera commented on mr/src/main/java/org/elasticsearch/hadoop/rest/query/SimpleQueryParser.java:
```diff
@@ -87,7 +87,7 @@ static void applyURIParameter(QueryStringQueryBuilder builder, String name, Stri
             builder.lenient(Boolean.parseBoolean(value.trim()));
         } else {
-            throw new IllegalArgumentException("Unknown " + name + " parameter; please change the URI query into a Query DLS (see 'Query String Query')");
+            throw new IllegalArgumentException("Unknown " + name + " parameter; please change the URI query into a Query DSL (see 'Query String Query')");
```
nice

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsStorageHandler.java:
```diff
@@ -18,36 +18,39 @@
+import java.util.Map;
+import java.util.Properties;
+
+import static org.elasticsearch.hadoop.hive.HiveConstants.*;
```
Re-expand this import into the explicit types.

AbelChan1989 replied:
done

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/Utils.java:
```diff
@@ -0,0 +1,85 @@
+import org.elasticsearch.hadoop.hive.EsStoragePredicateHandler;
+import org.elasticsearch.hadoop.util.unit.Booleans;
+
+public class Utils {
```
I think we should avoid adding a "Utils" class here. Maybe these constants and methods can be moved to different places?

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/Utils.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * sargable parser class path, default org.elasticsearch.hadoop.hive.EsStoragePredicateHandler
+     */
+    static String DATA_SOURCE_SARGABLE_CLASS_PATH = "es.internal.hive.sargable.class";
```
These config fields could be added to `org.elasticsearch.hadoop.hive.HiveConstants`

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/Utils.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * pushdown class path，default org.elasticsearch.hadoop.hive.EsStoragePredicateHandler
+     */
+    static String DATA_SOURCE_PUSH_DOWN_CLASS = "es.internal.hive.pushdown.class";
```
Are these actually internal configurations? The `es.internal` namespace is for properties that users are never expected to set themselves, and are reserved for use by the connector code only. If that's not the case for these, I would remove the `internal` part from the config key and prefix it with `ES_`.

If they _are_ meant to be internal properties, then their property names should be prefixed with `ES_INTERNAL` for easier code reading.

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/Utils.java:
```diff
@@ -0,0 +1,85 @@
+     */
+    static String DATA_SOURCE_SARGABLE_CLASS_PATH = "es.internal.hive.sargable.class";
+
+    public static boolean isPushDown(Settings cfg) {
```
This method could move to `org.elasticsearch.hadoop.hive.HiveUtils`

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/Utils.java:
```diff
@@ -0,0 +1,85 @@
+    private static final Log log = LogFactory.getLog(Utils.class);
+
+    /**
+     * is pushdown optimization, default true
```
For all of these properties that have default values - Can you add a default value static variable for each?

`static boolean DATA_SOURCE_PUSH_DOWN_DEFAULT = true; // for instance`

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/Utils.java:
```diff
@@ -0,0 +1,85 @@
+        return isPushdown;
+    }
+
+    public static SargableParser getSargableParser(Settings cfg) {
```
Is there a dire need to have this constructed by reflection? I'm not sure I see any cases where this implementation would be swapped out.

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/Utils.java:
```diff
@@ -0,0 +1,85 @@
+        return null;
+    }
+
+    public static PredicateHandler getPredicateHander(Settings cfg) {
```
Same here: Not sure we need to construct this by reflection. Open to discussion on it if you think there are good reasons for it.

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/util/HdfsUtil.java:
```diff
@@ -0,0 +1,58 @@
+     * @return
+     * @throws IOException
+     */
+    public static String getFileContent(String hdfsPath, Configuration conf) throws IOException {
```
Can we inline this into `org.elasticsearch.hadoop.hive.pushdown.parse.EsQueryParser#parseAsHdfsPath` instead of creating a Util class for it?

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/util/CollectionUtil.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class CollectionUtil {
+
+    public static <T> T safeget(List<T> list, int index) {
```
This could be moved to be an internal method on `TreeParser`

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsStorageHandler.java:
```diff
@@ -18,36 +18,39 @@
-import static org.elasticsearch.hadoop.hive.HiveConstants.COLUMNS;
-import static org.elasticsearch.hadoop.hive.HiveConstants.COLUMNS_TYPES;
-import static org.elasticsearch.hadoop.hive.HiveConstants.TABLE_LOCATION;
+import java.util.Map;
```
done

jbaiera commented:
Relates #4

AbelChan1989 commented:
HI, @jbaiera ,
  I am so Sorry for the delay on feedback.Corresponding changes have been made to the comments you have made. The list of changes is as follows:

- The CollectionUtil#safeget method moved to be an internal method on TreeParser and delete the CollectionUtil class.
- The HdfsUtil#getFileContent method moved to be the parseAsHdfsPath internal method on EsQueryParser and delete the HdfsUtil class.
- The JsonObjUtil class moved to org.elasticsearch.hadoop.hive.pushdown.parse.query package and renamed JsonObjManager, then delete the class.
- Fixed NullPointer exception when creating the PredicateHandler object.
- Reconstruct the field name mapping of the pushdown module to use FiledAlias directly.
- Fix bug for exists and missing grammars.
- Optimize package import, remove unnecessary '*' imports.
- Change the mode of creating EsStoragePredicateHandler and EsSargableParser objects to singleton.
- Add testcase for pushdown module.

Please review again. Thank you very much.

Thanks for a great product.

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/HiveTreeBuilder.java:
```diff
@@ -0,0 +1,167 @@
+ */
+public class HiveTreeBuilder {
+
+    protected static Log log = LogFactory.getLog(HiveTreeBuilder.class);
```
Nit: Can be `private static final`

AbelChan1989 replied:
done

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/parse/EsQueryParser.java:
```diff
@@ -0,0 +1,80 @@
+    protected final QueryBuilder rawQuery;
+    protected final boolean isES50;
+
+    public EsQueryParser(QueryBuilder queryBuilder, boolean isES50) {
```
The only place that this is used is parsing the contents of the `es.query` property to be combined with the pushdown query, but the internal pushdown filter setting should make this class no longer needed.

AbelChan1989 replied:
Yes, this class has been del .

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/parse/EsTreeParser.java:
```diff
@@ -0,0 +1,272 @@
+import java.util.Set;
+
+public class EsTreeParser extends TreeParser<JsonObj> {
+    JsonObj preFilterJson = null;
```
Not really needed. See other comments

AbelChan1989 replied:
Yes, del yet.

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/parse/EsTreeParser.java:
```diff
@@ -0,0 +1,272 @@
+public class EsTreeParser extends TreeParser<JsonObj> {
+    JsonObj preFilterJson = null;
+    List<JsonObj> queries = new LinkedList<JsonObj>();
+    final boolean isES50;
```
Nit: Make this `private final`

AbelChan1989 replied:
done

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/parse/EsTreeParser.java:
```diff
@@ -0,0 +1,272 @@
+
+public class EsTreeParser extends TreeParser<JsonObj> {
+    JsonObj preFilterJson = null;
+    List<JsonObj> queries = new LinkedList<JsonObj>();
```
This variable never seems to be updated with any contents?

AbelChan1989 replied:
del yet

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/HiveTreeBuilder.java:
```diff
@@ -0,0 +1,167 @@
+        return root;
+    }
+
+    private void _build(Node nowParent, ExprNodeDesc hiveNode) {
```
Nit: Standard java method naming please, I'd use something like `doBuild` if you want to differentiate from above `build` method.

AbelChan1989 replied:
done. method name changed from _build to doBuild

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/TreeParser.java:
```diff
@@ -0,0 +1,110 @@
+        return _parse(root);
+    }
+
+    protected T _parse(OpNode opNode) {
```
Nit: Follow java method naming conventions here. Could use `doParse`.

AbelChan1989 replied:
done. method name changed from _parse to doParse

jbaiera commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/TreeParser.java:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @param <T>
+ */
+public abstract class TreeParser<T> implements Serializable {
```
The logic for this is split across two classes, but I don't see any situations where we would need to use a different subclass of TreeParser. Can we simplify here?

AbelChan1989 replied:
done, same reason as above

jbaiera commented:
@AbelChan1989 Just checking in, are you still interested in continuing with moving this PR forward?

elasticcla commented:
Hi @AbelChan1989, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch-hadoop/pull/1111.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsStoragePredicateHandler.java:
```diff
@@ -0,0 +1,118 @@
+/**
+ * EsStoragePredicateHandler is an entrance class for implementing pushdown optimization.
+ */
+public class EsStoragePredicateHandler extends PredicateHandler<JsonObj> {
```
Because our company needs to connect to other data sources, such as hbase. Current, It has been changed to one class, and del 'PredicateHandler' class .

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/PredicateHandler.java:
```diff
@@ -0,0 +1,119 @@
+            return null;
+    }
+
+    public String serializePPD(T obj) {
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/PredicateHandler.java:
```diff
@@ -0,0 +1,119 @@
+
+    public abstract Pair<Node, T> optimizePushdown(JobConf jobConf, Deserializer deserializer, ExprNodeDesc exprNodeDesc);
+
+    public ExprNodeGenericFuncDesc translate2ExprNodeGenericFuncDesc(ExprNodeDesc desc) {
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsStoragePredicateHandler.java:
```diff
@@ -0,0 +1,118 @@
+            //set es mapping name
+            Map<String, String> mappingNames = new HashMap<String, String>();
+
+            String[] split = esMappingNameProps.split(",");
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/HiveTreeBuilder.java:
```diff
@@ -0,0 +1,169 @@
+            return null;
+
+        String[] arr = udfName.split("\\.");
+        udfName = arr[arr.length - 1];
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/EsSargableParser.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Semantic conversion from hive udf to es
+     */
+    protected BiMap<String, String> sargableOpUDFClassMapping = HashBiMap.create();
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/parse/EsTreeParser.java:
```diff
@@ -0,0 +1,258 @@
+
+import java.util.*;
+
+public class EsTreeParser extends TreeParser<JsonObj> {
```
Because our company needs to connect to other data sources, such as hbase. Current, It has been changed to one class, and del 'TreeParser' class .

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/PredicateHandler.java:
```diff
@@ -0,0 +1,119 @@
+    public abstract Pair<Node, T> optimizePushdown(JobConf jobConf, Deserializer deserializer, ExprNodeDesc exprNodeDesc);
+
+    public ExprNodeGenericFuncDesc translate2ExprNodeGenericFuncDesc(ExprNodeDesc desc) {
+        if (desc instanceof ExprNodeGenericFuncDesc)
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/PredicateHandler.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    public String serializePPD(T obj) {
+        if (obj != null)
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/PredicateHandler.java:
```diff
@@ -0,0 +1,119 @@
+
+    @Override
+    public DecomposedPredicate decomposePredicate(JobConf jobConf, Deserializer deserializer, ExprNodeDesc exprNodeDesc) {
+        if (exprNodeDesc == null)
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsHiveInputFormat.java:
```diff
@@ -110,6 +113,17 @@ public Path getPath() {
+            String esQuery = Utilities.deserializeObject(filterObjectStr, String.class);
+            if (esQuery != null && esQuery.length() > 0) {
+                log.info("[EsHiveInputFormat][Pushdown][Es.query] : " + esQuery);
+                settings.setQuery(esQuery);
```
Ok, I didn't find this  method before, crying.  Currently it has adjusted to call the settingUtils.setFilters method, then combined when scroll begin to search, please check!

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/EsSargableParser.java:
```diff
@@ -0,0 +1,142 @@
+ */
+public class EsSargableParser implements SargableParser {
+
+    private static EsSargableParser _instance;
```
done. Firstly, remove the singleton to make it become a normal class. Then, the map and set is adjusted to static fields.

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsStoragePredicateHandler.java:
```diff
@@ -0,0 +1,123 @@
+
+    private static Log log = LogFactory.getLog(EsStoragePredicateHandler.class);
+
+    private static EsStoragePredicateHandler _instance;
```
done. remove the singleton to make it become a normal class

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/HiveConstants.java:
```diff
@@ -18,14 +18,16 @@
  */
 package org.elasticsearch.hadoop.hive;
 
+import org.elasticsearch.hadoop.hive.pushdown.EsSargableParser;
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsStoragePredicateHandler.java:
```diff
@@ -0,0 +1,123 @@
+        log.info("[PushDown][isES50] : " + isES50);
+
+        SargableParser sargableParser = EsSargableParser.getInstance();
+        if (sargableParser == null) {
```
Remove singleton mode and it is impossible to null.

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/HiveTreeBuilder.java:
```diff
@@ -0,0 +1,167 @@
+
+    protected static Log log = LogFactory.getLog(HiveTreeBuilder.class);
+
+    protected SargableParser sargableParser = null;
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/pushdown/HiveTreeBuilder.java:
```diff
@@ -0,0 +1,167 @@
+    /**
+     * load es.mapping,names ,then get the field mmaping of hive to elasticsearch
+     */
+    protected FieldAlias fieldAlias = null;
```
done

AbelChan1989 commented on hive/src/main/java/org/elasticsearch/hadoop/hive/EsStoragePredicateHandler.java:
```diff
@@ -0,0 +1,123 @@
+        EsTreeParser parser = new EsTreeParser(sargableParser, isES50);
+
+        String preFilterQuery = settings.getQuery();
+        // if exists es.query prop, then it is a necessary condition to add to the pushdown optimization plan
```
done.

AbelChan1989 commented:
HI ,@jbaiera. I am soooo sorry that I haven't time to deal with this PR for personal reasons last year. The PR will be actively followed up from current.

Additionally, this CI error seems to have nothing to do with the code I submitted, please check.

Thanks

jbaiera commented:
@AbelChan1989 It's totally fine! Glad that you are still interested in pushing this forward.

Unfortunately, CI is in a bit of a bad place right now and I'm working to get it running smoothly again across the ES project and this one. I'll try to get some time to review this soon.

AbelChan1989 commented:
> @AbelChan1989 It's totally fine! Glad that you are still interested in pushing this forward.
> 
> Unfortunately, CI is in a bit of a bad place right now and I'm working to get it running smoothly again across the ES project and this one. I'll try to get some time to review this soon.

ok, thanks.

I will continue to pay attention to this PR. (^_^)

pmogren commented:
With all respect, I'm just a passer-by looking for a potential solution, sad to see this work done and not merged for so long. I hope somebody can see it through soon. Maybe it's just been lost in the shuffle, and maybe this helps to surface it 🤞
@jbaiera @AbelChan1989

jbaiera commented:
Closing this out for now since there hasn't been much further movement on it. I would love to revisit this feature set in the future, but for the time being, I think it's best to close this out.
